crates/yulgen/src/runtime/mod.rs

```diff
@@ -1,174 +1,246 @@
             .collect::<Vec<_>>();
 
-        let mut funcs = [
-            std,
-            encoding,
-            decoding,
-            contract_calls,
-            revert_calls_from_assert,
-            revert_calls,
-            struct_apis,
+        let revert_panic_batch = vec![vec![AbiType::Uint { size: 32 }]];
+
+        let structs_batch: Vec<Vec<AbiType>> = contract
+            .module(db)
+            .all_structs(db)
+            .iter()
+            .map(|struc| vec![struc.typ(db).as_abi_type(db)])
+            .collect::<Vec<Vec<_>>>();
+
+        let batch = [
+            public_functions_batch,
+            events_batch,
+            contracts_batch,
+            assert_strings_batch,
+            revert_errors_batch,
+            revert_panic_batch,
+            structs_batch,
         ]
         .concat();
-        funcs.sort();
-        funcs.dedup();
-        return funcs;
-    }
-
-    panic!("missing contract attributes")
-}
+        functions::abi::batch_encode(batch)
+    };
+    let decoding = {
+        let public_functions_batch = public_functions
+            .iter()
+            .map(|sig| {
+                (
+                    to_abi_types(db, &sig.param_types()),
+                    AbiDecodeLocation::Calldata,
+                )
+            })
+            .collect();
+
+        let init_params_batch = if let Some(init_fn) = contract.init_function(db) {
+            let sig = init_fn.signature(db);
+            vec![(
+                to_abi_types(db, &sig.param_types()),
+                AbiDecodeLocation::Memory,
+            )]
+        } else {
+            vec![]
+        };
 
-/// Concatenates the functions inside of each contract.
-fn concat_contract_functions(contracts: Vec<Contract>) -> Vec<FunctionAttributes> {
-    contracts
-        .into_iter()
-        .map(|contract| contract.functions)
+        let contracts_batch = external_functions
+            .iter()
+            .filter_map(|function| {
+                let return_type = function.expect_return_type();
+                (!return_type.is_unit())
+                    .then(|| (vec![return_type.as_abi_type(db)], AbiDecodeLocation::Memory))
+            })
+            .collect();
+
+        let batch = [public_functions_batch, init_params_batch, contracts_batch].concat();
+        functions::abi::batch_decode(batch)
+    };
+    let contract_calls = {
+        external_contracts
+            .iter()
+            .map(|contract| functions::contracts::calls(db, contract.to_owned()))
+            .collect::<Vec<_>>()
+            .concat()
+    };
+
+    let struct_apis = contract
+        .module(db)
+        .all_structs(db)
+        .iter()
+        .map(|struct_| {
+            let fields = struct_
+                .fields(db)
+                .iter()
+                .map(|(name, id)| {
+                    (
+                        name.to_string(),
+                        id.typ(db).expect("struct field type error").as_abi_type(db),
```

Comment:
To continue with the other comment.. we can just leave the call to `as_abi_type` out and pass the `FixedSize` types into the `struct_apis` function, right?